flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -177,7 +193,8 @@ public void recycle(MemorySegment segment) {
 
 		final List<MemorySegment> segments = new ArrayList<>(numberOfSegmentsToRequest);
 		try {
-			while (segments.size() < numberOfSegmentsToRequest) {
+			final Deadline deadline = Deadline.now().plus(Duration.ofMillis(requestSegmentsTimeoutInMillis));
```

Comment:
This is probably a system call (which might be quite costly), could we restructure the code in a such way, that we make this only after first timeout to `availableMemorySegments.poll(2, TimeUnit.SECONDS);`?

Or even remove it entirely: if `poll(2, TimeUnit.SECONDS)` returns `null` maybe we can assume that 2 seconds have passed and subtract that from the timeout time?

Reply:
I would favor simplicity/robustness here. It is a rare call after all.

Reply:
Very thanks for all the comments @pnowojski and @StephanEwen. for the two options to optimize the performance, I prefer to the second option since the number of exclusive buffers might be larger than 1 and factoring out the first `poll` might not reduce the overhead for normal cases. 

For the second one, to not increase much complexity we may reuse the operations of Duration like 

<pre>
//private static final Duration SEGMENT_POLL_TIME_ELAPSED = Duration.ofSeconds(2);

Duration remainTimes = Duration.ofNanos(requestSegmentsTimeout.toNanos());

while (true) {
	if (isDestroyed) {
		throw new IllegalStateException("Buffer pool is destroyed.");
	}

	final MemorySegment segment = availableMemorySegments.poll(2, TimeUnit.SECONDS);
	if (segment != null) {
		segments.add(segment);
	}

	if (segments.size() >= numberOfSegmentsToRequest) {
		break;
	}

	remainTimes = remainTimes.minus(SEGMENT_POLL_TIME_ELAPSED);

	if (remainTimes.toNanos() <= 0) {
		throw new IOException(String.format("Timeout triggered when requesting exclusive buffers: %s, " +
						" or you may increase the timeout which is %dms by setting the key '%s'.",
				getConfigDescription(),
				requestSegmentsTimeout.toMillis(),
				NettyShuffleEnvironmentOptions.NETWORK_EXCLUSIVE_BUFFERS_REQUEST_TIMEOUT_MILLISECONDS.key()));
	}
}
</pre>

On the other side, indeed the method is only called when creating the single input channels during task creation and should not affect the buffer transmission/receive. Reusing `Deadline` should be able to hide the computation logic of the timeout. So do we prefer to the current `Deadline` method or the above `Duration` method ?

Reply:
Let's stick with the `Deadline` as in your proposal there is already a small bug (for the timeout based on subtraction to work, you would have to subtract it only if `segment == null`, otherwise even if get the buffer after only a 1 millisecond of waiting, you subtract 2 seconds of waiting time).